File: server/src/ai/prompts.js
/**
 * Structured prompts for AI-powered payment diagnosis.
 * Each prompt is designed to extract structured JSON from the LLM
 * with failure classification, recoverability scoring, and strategy selection.
 */

const SYSTEM_PROMPT = `You are an expert payment recovery analyst working for an Indian fintech platform. 
You analyze failed payment transactions and diagnose the root cause with high accuracy.
You understand Indian banking infrastructure: UPI mandates, NACH/eNACH, RBI regulations on auto-debit, 
card network behaviors (Visa/Mastercard/RuPay), and common failure patterns across Indian issuing banks.

Your responses must ALWAYS be valid JSON matching the exact schema requested.
Never include explanatory text outside the JSON object.`;

function buildDiagnosisPrompt(transaction) {
  const { amount, type, errorCode, bank, cardBin, timestamp, customerInfo } = transaction;

  return {
    system: SYSTEM_PROMPT,
    user: `Analyze this failed payment transaction and provide a diagnosis:

TRANSACTION DETAILS:
- Amount: ₹${amount}
- Type: ${type} (${type === 'subscription' ? 'recurring charge' : 'one-time payment'})
- Failure Code: ${errorCode || 'UNKNOWN'}
- Issuing Bank: ${bank}
- Card BIN: ${cardBin}
- Timestamp: ${timestamp}
- Customer Type: ${customerInfo?.historySummary || 'UNKNOWN'}
- Previous Failed Attempts: ${customerInfo?.previousAttempts || 0}

Respond with this exact JSON structure:
{
  "failureReason": "one of: CARD_EXPIRED | INSUFFICIENT_FUNDS | BANK_DECLINE | NETWORK_ERROR | FRAUD_SUSPECTED | MANDATE_REVOKED | CHECKOUT_FRICTION | PRICE_SHOCK",
  "recoverabilityScore": <number 0-100, where 100 = highly recoverable>,
  "reasoning": "<2-3 sentence human-readable explanation of why you classified it this way and what signals you used>",
  "suggestedStrategy": "one of: SMART_RETRY | NUDGE_UPDATE | ALT_METHOD | IMMEDIATE_RETRY | ESCALATE | GENTLE_NUDGE | SIMPLIFY_RETRY | VALUE_REMIND",
  "optimalRetryWindow": "<ISO timestamp for best retry time, or null if not applicable>",
  "confidence": <number 0-100>,
  "riskFactors": ["<list of risk factors that could prevent recovery>"]
}`
  };
}

function buildDegradationPrompt(degradationData) {
  return {
    system: SYSTEM_PROMPT,
    user: `Analyze this payment degradation pattern:

DEGRADATION DETAILS:
- Dimension: ${degradationData.dimension} = ${degradationData.dimensionValue}
- Current Success Rate: ${(degradationData.successRate * 100).toFixed(1)}%
- Sample Size: ${degradationData.sampleSize} transactions
- Total Failed Amount: ₹${degradationData.failedAmount}

Respond with this exact JSON structure:
{
  "rootCause": "<most likely root cause of the degradation>",
  "severity": "one of: LOW | MEDIUM | HIGH | CRITICAL",
  "suggestedAction": "one of: ROUTE_SWITCH | RETRY_TIMING | ALERT_MERCHANT | MONITOR | ESCALATE",
  "reasoning": "<2-3 sentence explanation>",
  "estimatedRecoveryTime": "<estimated time for the issue to resolve, e.g., '2 hours', '1 day'>",
  "confidence": <number 0-100>
}`
  };
}

function buildNudgePrompt(transaction, strategy) {
  return {
    system: `You are a customer communication specialist for an Indian payments platform. 
Write empathetic, concise recovery messages in a professional yet friendly tone. 
The message should feel personal, not automated. Use Hinglish sparingly if the customer context suggests it.
Keep SMS messages under 160 characters. Keep email subjects under 60 characters.`,
    user: `Generate a recovery nudge for this transaction:

CONTEXT:
- Customer Type: ${transaction.customerInfo?.historySummary || 'RETURNING_CUSTOMER'}
- Amount: ₹${transaction.amount}
- Failure Reason: ${transaction.errorCode}
- Recovery Strategy: ${strategy}
- Previous Attempts: ${transaction.customerInfo?.previousAttempts || 0}

Respond with this exact JSON structure:
{
  "sms": "<SMS message, max 160 chars>",
  "emailSubject": "<email subject, max 60 chars>",
  "emailBody": "<email body, 2-3 sentences>",
  "tone": "one of: URGENT | FRIENDLY | EMPATHETIC | PROFESSIONAL",
  "includesIncentive": <boolean>,
  "incentiveDetails": "<if applicable, describe the incentive>"
}`
  };
}

module.exports = {
  SYSTEM_PROMPT,
  buildDiagnosisPrompt,
  buildDegradationPrompt,
  buildNudgePrompt,
};

File: server/app/ai/prompts.py
DIAGNOSTIC_SYSTEM_PROMPT = """You are the X.A.V.I.E.R. agent, an expert payment failure diagnostician for the Indian digital payments ecosystem.

You have access to a curated Knowledge Base (KB) of documented payment failure patterns with historical recovery data. Use the provided KB evidence to ground your diagnosis. Do NOT ignore the KB entries — they contain real-world success rates and RBI compliance guidelines.

Your task:
1. Read the transaction data carefully
2. Study the matching Knowledge Base entries provided
3. Combine the KB evidence with your own reasoning to produce a diagnosis
4. If the KB recommends a strategy, prefer it unless you have strong reason to deviate

You MUST respond in valid JSON format matching this exact schema:
{
  "reason": "Clear, specific explanation of why this payment failed",
  "confidence": 0-100 (integer, higher if KB match is strong),
  "strategy": "One of: SUBSCRIPTION_RECOVERY, PAYMENT_DEGRADATION, CHECKOUT_DROPOFF",
  "riskFactors": ["list", "of", "contributing", "factors"],
  "recommendedAction": "Specific, actionable next step to recover this payment",
  "optimalRetryWindow": "immediate, 1h, 4h, 24h, or none",
  "knowledgeBaseMatch": "The KB entry ID that most closely matches (e.g., KB-001), or null if no match"
}

Strategy Selection Guidelines (from KB patterns):
- INSUFFICIENT_FUNDS, RISK_REJECT, EXPIRED_CARD, mandate/subscription failures -> SUBSCRIPTION_RECOVERY
- GATEWAY_TIMEOUT, PROCESSOR_DOWN, NETWORK_ERROR, routing/settlement failures -> PAYMENT_DEGRADATION
- USER_ABANDONED, 3DS_FAILED, OTP_TIMEOUT, PAGE_CRASH, session/checkout friction -> CHECKOUT_DROPOFF
"""

def generate_diagnostic_prompt(transaction, kb_context=""):
    """
    Build the user prompt for diagnosis.
    Now accepts an optional kb_context string containing formatted KB entries.
    """
    base = f"""## Transaction Data
- Transaction ID: {transaction['id']}
- Amount: ₹{transaction['amount']} {transaction.get('currency', 'INR')}
- Gateway: {transaction.get('gateway', 'Unknown')}
- Bank: {transaction.get('bankName', 'Unknown')}
- Payment Method: {transaction.get('paymentMethod', 'Unknown')}
- Customer Is Repeat: {transaction.get('customerInfo', {}).get('isRepeat', False)}
- Previous Failed Attempts: {transaction.get('customerInfo', {}).get('previousAttempts', 0)}
- Device: {transaction.get('metadata', {}).get('device', 'Unknown')}

## Error Details
- Error Code: {transaction.get('errorDetails', {}).get('code', 'N/A')}
- Error Reason: {transaction.get('errorDetails', {}).get('reason', 'N/A')}
- Error Message: {transaction.get('errorDetails', {}).get('message', 'N/A')}"""

    if kb_context:
        base += f"""

## Relevant Knowledge Base Entries
The following are the most relevant historical failure patterns from our curated database. Use these as evidence to inform your diagnosis:

{kb_context}"""

    base += """

## Task
Analyze this payment failure using both the transaction data and the knowledge base evidence above. Return your JSON diagnosis."""

    return base

NUDGE_SYSTEM_PROMPT = """You are a highly empathetic, high-converting customer success agent.
Write a personalized SMS/Email nudge to recover an abandoned or failed checkout.
Keep it under 160 characters if possible. Be helpful, not pushy.

You MUST respond in valid JSON format matching this schema:
{
  "content": "The actual message text",
  "tone": "empathetic, urgent, or helpful",
  "channel": "sms or email"
}"""

def generate_nudge_prompt(context):
    return f"""Customer ID: {context.get('customer_id')}
Failure Reason: {context.get('reason')}
Amount: {context.get('amount')}
Context: {context.get('context_string')}

Write a recovery nudge for this specific situation."""
